libp2p/src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt

```diff
@@ -408,11 +412,28 @@ open class GossipRouter(
                     .plus(getDirectPeers())
             } else {
                 msg.topics
-                    .mapNotNull { topic ->
-                        mesh[topic] ?: fanout[topic] ?: getTopicPeers(topic).shuffled(random).take(params.D)
-                            .also {
-                                if (it.isNotEmpty()) fanout[topic] = it.toMutableSet()
-                            }
+                    .map { topic ->
+                        val topicMeshPeers = mesh[topic]
+                        if (topicMeshPeers != null) {
+                            // we are subscribed to the topic
+                            val nonMeshTopicPeers = getTopicPeers(topic) - topicMeshPeers
+                            val (nonMeshTopicPeersAbovePublishThreshold, nonMeshTopicPeersBelowPublishThreshold) =
+                                nonMeshTopicPeers.partition { score.score(it.peerId) >= scoreParams.publishThreshold }
+                            // this deviates from the original spec but we want at least D peers for publishing
+                            // prioritizing mesh peers, then non-mesh peers with acceptable score,
+                            // and then underscored non-mesh peers as a last resort
+                            (
+                                topicMeshPeers +
+                                    nonMeshTopicPeersAbovePublishThreshold.shuffled(random) +
+                                    nonMeshTopicPeersBelowPublishThreshold.shuffled(random)
+                                ).take(params.D)
```

Comment:
the logic makes sense to me, but I don't like that we and up always doing the partitioning and shuffling even if the topic fulfill the `D` param (which is supposed to be the most common case). Maybe we can go for it only when topicPeers is < `D`?

Reply:
Yep, just made it alongside with fixing a regression